src/cpu.c

```diff
@@ -21,55 +22,70 @@ cpu* mgec_new_cpu() {
   return c;
 }
 
-#define load(C, R, V) _Generic((V), u8: load8, u16: load16)(C, R, V)
-
-static inline cycles load8(cpu* c, cpu_register_t reg, u8 value) {
+inline cycles load8(cpu* c, cpu_register_t reg, u8 value) {
+  // TODO: flag changes
+  // zero out the high byte when in native mode or
+  // emulation mode with 8-bit indexers
+  bool hbo = (c->e || (c->sr_x || c->sr_m));
   switch (reg) {
   case XL:
     c->xl = value;
-    break;
-  case XH:
-    c->xh = value;
+    if (hbo)
+      c->xh = 0;
     break;
   case YL:
     c->yl = value;
+    if (hbo)
+      c->yh = 0;
     break;
-  case YH:
-    c->yh = value;
-    break;
-  // Making sure we're handling all cases
   // TODO: some error condition
   case X:
+  case XH:
   case Y:
+  case YH:
   case A:
   case B:
   case C:
   default:
     break;
   }
 
+  if (value == 0) {
```

Comment:
These look good, now you just need to check if bit 7 (the most significant bit) is set to set the negative flag.